File: 승준/20230627/binary_search_tree_5639.java
import java.io.*;
import java.util.*;

public class binary_search_tree_5639 {
    static class Node {
        int num;
        Node left, right;

        Node(int num) {
            this.num = num;
        }

        Node(int num, Node left, Node right) {
            this.num = num;
            this.left = left;
            this.right = right;
        }

        void insert(int nodeNum) {
            if (nodeNum < this.num) {
                if (this.left == null) {
                    this.left = new Node(nodeNum);
                } else {
                    this.left.insert(nodeNum);
                }
            } else {
                if (nodeNum > this.num) {
                    if (this.right == null) {
                        this.right = new Node(nodeNum);
                    } else {
                        this.right.insert(nodeNum);
                    }
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        Node root = new Node(Integer.parseInt(br.readLine()));
        String input;
        while (true) {
            input = br.readLine();
            if (input == null || input.equals("")) {
                break;
            }
            root.insert(Integer.parseInt(input));
        }

        postOrder(root);
    }

    static void postOrder(Node node) {
        if (node == null) {
            return;
        }

        postOrder(node.left);
        postOrder(node.right);
        System.out.println(node.num);
    }
}

File: 승준/20230626/binary_search_tree_5639.py
def get_postorder(preorder):
    if not preorder:
        return []

    root = preorder[0]
    left = []
    right = []
    
    for node in preorder[1:]:
        if node < root:
            left.append(node)
        else:
            right.append(node)
    
    left_postorder = get_postorder(left)
    right_postorder = get_postorder(right)
    
    return left_postorder + right_postorder + [root]


# 전위 순회
preorder= []
while True:
    try:
        node = int(input())
        preorder.append(node)
    except EOFError:
        break
    
postorder = get_postorder(preorder)
for n in postorder:
    print(n)